pydomcfg/domzgr/zco.py

```diff
@@ -59,216 +43,268 @@ def __call__(
 
-        Parameters
-        ----------
-        ds: Dataset
-            xarray dataset with empty ``e3{T,W}`` and ``z3{T,W}``
-            correctly computed
-        Returns
-        -------
-        ds: Dataset
-            xarray dataset with ``e3{T,W}`` correctly computed
+        # Return here if is uniform
+        if self._is_uniform:
+            # 0d DataArray
+            e3 = DataArray((self._pphmax / (self._jpk - 1.0)))
+            return tuple([e3, e3])
+
+        # Stretched zco grid
+        zco_e3 = []
+        for sigma in self._sigmas:
+            a0 = self._ppa0
+            a1 = self._ppa1
+            kk = -sigma * (self._jpk - 1.0) + 1.0
+            tanh1 = np.tanh((kk - self._ppkth) / self._ppacr)
+
+            # Compute e3
+            e3 = a0 + a1 * tanh1
+
+            # Add double tanh term
+            if self._ldbletanh:
+                a2 = self._ppa2
+                tanh2 = np.tanh((kk - self._ppkth2) / self._ppacr2)
+                e3 += a2 * tanh2
+
+            # Append
+            zco_e3 += [e3]
+
+        return tuple(zco_e3)
+
+    @staticmethod
+    def _get_ldbletanh_and_pp2(
+        ldbletanh: Optional[bool], pp2: Tuple[Optional[float], ...]
+    ) -> Tuple[bool, Tuple[float, ...]]:
+        """
+        If ldbletanh is None, its bool value is inferred from pp2.
+        Return pp2=(0, 0, 0) when double tanh is switched off.
         """
 
-        sigmas = [self._sigT, self._sigW]
+        pp_are_none = tuple(pp is None for pp in pp2)
+        prefix_msg = "ppa2, ppkth2 and ppacr2"
 
-        for varname, sig in zip(["e3T", "e3W"], sigmas):
-            if self._is_uniform:
-                # uniform zco grid
-                ds[varname][{"z": slice(self._jpk)}] = self._pphmax / (self._jpk - 1.0)
-            else:
-                # stretched zco grid
-                a0 = self._ppa0
-                a1 = self._ppa1
-                # now faster to use loop, to be optimised
-                # using xarray features in the future
-                for k in range(self._jpk):
-                    kk = -sig[{"z": k}] * (self._jpk - 1.0) + 1.0
-                    tanh1 = np.tanh((kk - self._ppkth) / self._ppacr)
-                    if self._ldbletanh:
-                        a2 = self._ppa2
-                        tanh2 = np.tanh((kk - self._ppkth2) / self._ppacr2)
-                    else:
-                        a2 = tanh2 = 0.0
-
-                    ds[varname][{"z": k}] = a0 + a1 * tanh1 + a2 * tanh2
-
-        return ds
+        # Errors
+        if ldbletanh is True and any(pp_are_none):
+            raise ValueError(f"{prefix_msg} MUST be all float when ldbletanh is True")
+        if ldbletanh is None and (any(pp_are_none) and not all(pp_are_none)):
+            raise ValueError(f"{prefix_msg} MUST be all None or float")
+
+        # Warning
+        if ldbletanh is False and not all(pp_are_none):
+            warnings.warn(f"{prefix_msg} are ignored when ldbletanh is False")
```

Comment:
Does this look right? Before we were doing a more strict check (pp>0 rather than not null), but I think it makes more sense to default None and the user can fully play with the underlying equation. If you think we should restore pp>0 when ldbletanh is True, then maybe we should introduce more checks on all parameters (e.g., ppdzmin >= 0, ...)